src/zenml/utils/run_utils.py

```diff
@@ -72,3 +73,73 @@ def stop_run(run: PipelineRunResponse, graceful: bool = False) -> None:
 
     # Stop the run
     orchestrator.stop_run(run=run, graceful=graceful)
+
+
+def refresh_run_status(
+    run: PipelineRunResponse,
+    include_step_updates: bool = False,
+) -> PipelineRunResponse:
+    """Refresh the status of a pipeline run if it is initializing/running.
+
+    Args:
+        run: The pipeline run to refresh.
+        include_step_updates: Flag deciding whether we should also refresh
+            the status of individual steps.
+
+    Returns:
+        The refreshed pipeline run.
+    """
+    # Check if the stack still accessible
+    if run.stack is None:
+        raise ValueError(
+            "The stack that this pipeline run response was executed on"
+            "is either not accessible or has been deleted."
+        )
+
+    # Check if the orchestrator is still accessible
+    orchestrator_list = run.stack.components.get(
+        StackComponentType.ORCHESTRATOR, []
+    )
+    if len(orchestrator_list) == 0:
+        raise ValueError(
+            "The orchestrator that this pipeline run response was "
+            "executed with is either not accessible or has been deleted."
+        )
+
+    orchestrator = cast(
+        BaseOrchestrator,
+        StackComponent.from_model(component_model=orchestrator_list[0]),
+    )
+
+    # Fetch the status (and optionally step statuses)
+    pipeline_status, step_statuses = orchestrator.fetch_status(
+        run=run, include_steps=include_step_updates
+    )
+
+    client = Client()
+
+    # Update step statuses
+    if include_step_updates:
+        assert step_statuses is not None
+
+        for step_name, step_status in step_statuses.items():
```

Comment:
I don't think this logic works. What if the step is `Retried` in the ZenML server, but that is not something that you can detect from the orchestrator? If a step is already in a finished state, it should IMO not be updated anymore.

Reply:
> If a step is already in a finished state, it should IMO not be updated anymore.

In the Kubernetes orchestrator, we enforce this in the `_fetch_step_status`. Do you think we should also enforce this here?

Reply:
Enforced here now.